Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPattern.cpp

```diff
@@ -0,0 +1,441 @@
+ * consist of three parts: scheme, host, and path.
+ * 
+ * Generally, match patterns are returned from a #WebKitWebExtension.
+ */
+struct _WebKitWebExtensionMatchPattern {
+#if ENABLE(WK_WEB_EXTENSIONS)
+    _WebKitWebExtensionMatchPattern(RefPtr<WebExtensionMatchPattern>&& matchPattern)
+        : matchPattern(WTFMove(matchPattern))
+    {
+    }
+
+    RefPtr<WebExtensionMatchPattern> matchPattern;
+    CString string { matchPattern->string().utf8() };
+    CString scheme { matchPattern->scheme().utf8() };
+    CString host { matchPattern->host().utf8() };
+    CString path { matchPattern->path().utf8() };
+    bool matchesAllURLs { matchPattern->matchesAllURLs() };
+    bool matchesAllHosts { matchPattern->matchesAllHosts() };
+    int referenceCount { 1 };
+#else
+    _WebKitWebExtensionMatchPattern()
+    {
+    }
+#endif
+};
+
+G_DEFINE_BOXED_TYPE(WebKitWebExtensionMatchPattern, webkit_web_extension_match_pattern, webkit_web_extension_match_pattern_ref, webkit_web_extension_match_pattern_unref)
+
+#if ENABLE(WK_WEB_EXTENSIONS)
+
+/**
+ * webkit_web_extension_match_pattern_register_custom_URL_scheme:
+ * @urlScheme: The custom URL scheme to register
+ *
+ * Registers a custom URL scheme that can be used in match patterns.
+ * 
+ * This method should be used to register any custom URL schemes used by the app for the extension base URLs,
+ * other than `webkit-extension`, or if extensions should have access to other supported URL schemes when using `<all_urls>`.
+ */
+void webkit_web_extension_match_pattern_register_custom_URL_scheme(const gchar* urlScheme)
+{
+    g_return_if_fail(WTF::URLParser::maybeCanonicalizeScheme(String::fromUTF8(urlScheme)));
+
+    WebKit::WebExtensionMatchPattern::registerCustomURLScheme(String::fromUTF8(urlScheme));
+}
+
+WebKitWebExtensionMatchPattern* webkitWebExtensionMatchPatternCreate(const RefPtr<WebExtensionMatchPattern>& apiMatchPattern)
+{
+    if (!apiMatchPattern)
+        return nullptr;
+
+    ASSERT(API::Object::unwrap(static_cast<void*>(apiMatchPattern.get()))->type() == API::Object::Type::WebExtensionMatchPattern);
+    WebKitWebExtensionMatchPattern* matchPattern = static_cast<WebKitWebExtensionMatchPattern*>(fastMalloc(sizeof(WebKitWebExtensionMatchPattern)));
+    new (matchPattern) WebKitWebExtensionMatchPattern(static_pointer_cast<WebExtensionMatchPattern>(apiMatchPattern));
+    return matchPattern;
+}
+
+WebKitWebExtensionMatchPattern* webkitWebExtensionMatchPatternCreate(Ref<WebExtensionMatchPattern>& matchPattern)
+{
+    RefPtr<WebExtensionMatchPattern> apiMatchPattern = adoptRef(matchPattern.get());
+
+    WebKitWebExtensionMatchPattern* pattern = static_cast<WebKitWebExtensionMatchPattern*>(fastMalloc(sizeof(WebKitWebExtensionMatchPattern)));
+    new (pattern) WebKitWebExtensionMatchPattern(static_pointer_cast<WebExtensionMatchPattern>(apiMatchPattern));
+    return pattern;
+}
```

Comment:
Is it really necessary to have separate constructors for RefPtr and Ref?

If you do need to keep it, then at least implement the Ref constructor using the RefPtr constructor instead of copying all of this twice, to avoid inconsistencies. E.g. you have an assert in the RefPtr version that is missing here.

Reply:
Yes, WebExtensionMatchPattern itself returns RefPtr's, and WebExtension has some functions that return Ref, so when we expose a Gtk API for WebExtension, Ref will be needed.